File: benchmarks/reference/SortingBenchmarkReference/BinarySearchTree/Javascript/BST.js
function BSTNode(val, parent, tree) {
	this.value = val;
	this.parent = parent;
	this.tree = tree;
	this.left = null;
	this.right = null;
	
	this.add = function(elem) {
		if (elem > this.value) {
			if (this.right == null)
				this.right = new BSTNode(elem, this, tree);
			else
				this.right.add(elem);
		}
		else {
			if (this.left == null)
				this.left = new BSTNode(elem, this, tree);
			else
				this.left.add(elem);
		}
	};
	this.find = function(elem) {
		if (this.value == elem) return true;
		if (elem > this.value && this.right != null) {
			return this.right.search(elem);
		} else if (elem < this.value && this.left != null) {
			return this.left.search(elem);
		} else {
			return false;
		}
	};
	
	this.replaceParentChild = function(node) {
		if (this.parent != null) {
			if (this.value > this.parent.value) {
				this.parent.right = node;
			} else {
				this.parent.left = node;
			}
			if (node != null)
				node.parent = this.parent;
		}
		else {
			if (node != null)
				node.parent = null;
			this.tree.root = node;
		}
	}
	
	this.findDeleteMinNode = function() {
		if (this.left != null)
			return left.findDeleteMinNode();
		this.replaceParentChild(this.right);
		return this;
	}
	
	this.del = function(elem) {
		if (this.value != elem) {
			if (elem > this.value && this.right != null)
				this.right.del(elem);
			else if (elem < this.value && this.left != null)
				this.left.del(elem);
			return;
		}
		
		if (this.left != null && this.right != null) {
			this.value = this.right.findDeleteMinNode().value;
		}
		else if (this.left == null && this.right != null) {
			this.replaceParentChild(this.right);
		}
		else if (this.left != null && this.right == null) {
			this.replaceParentChild(this.left);
		}
		else if (this.left == null && this.right == null) {
			this.replaceParentChild(null);
		}
	};
}

function BST() {
	this.root = null;
	this.add = function(elem) {
		if (this.root != null)
			this.root.add(elem);
		else
			this.root = new BSTNode(elem, null, this);
	};
	this.find = function (elem) {
		if (this.root != null)
			return this.root.find(elem);
		return false;
	};
	this.del = function (elem) {
		if (this.root != null)
			this.root.del(elem);
	};
}

function benchmark(runner) {
	var bst = new BST();
	var n = 5000;
	var elems = new Array();
	for (var i = 0; i < n; i++) {
		var elem = Math.floor(Math.random() * (10000000 + 1));
		elems.push(elem);
	}
	runner.startTimer();
	for (var k = 0; k < n; k++) {
		bst.add(elems[i]);
	}
	for (var l = 0; l < n; l++) {
		bst.del(elems[i]);
	}
	runner.stopTimer();
}

File: benchmarks/reference/SortingBenchmarkReference/BinarySearchTree/Java/BST.java
package SortingBenchmarkReference.BinarySearchTree.Java;

public class BST {
    class BSTNode {
        public BSTNode(int value, BSTNode parent) {
            this.value = value;
			this.parent = parent;
        }

        BSTNode left;
        BSTNode right;
		BSTNode parent;
        int value;
		
		
		private void replaceParentChild(BSTNode node) {
            if (parent != null) {
                if (value > parent.value) {
                    parent.right = node;
                } else {
                    parent.left = node;
                }
                if (node != null)
                    node.parent = parent;
            }
            else {
                if (node != null)
                    node.parent = null;
                root = node;
            }
		}
        
        void add(int elem) {
            if (elem > value) {
                if (right == null)
                    right = new BSTNode(elem, this);
                else
                    right.add(elem);
            }
            else {
                if (left == null)
                    left = new BSTNode(elem, this);
                else
                    left.add(elem);
            }
        }

        boolean search(int elem) {
			if (value == elem) return true;
			if (elem > value && right != null) {
				return right.search(elem);
			} else if (elem < value && left != null) {
				return left.search(elem);
			} else {
				return false;
			}
		}

		private BSTNode findDeleteMinNode() {
			if (left != null)
				return left.findDeleteMinNode();
			replaceParentChild(this.right);
			return this;
		}

		void delete(int elem) {
			if (value != elem) {
				if (elem > value && right != null)
					right.delete(elem);
				else if (elem < value && left != null)
					left.delete(elem);
				return;
			}
			
			if (left != null && right != null) {
				this.value = right.findDeleteMinNode().value;
			}
			else if (left == null && right != null) {
				replaceParentChild(right);
			}
			else if (left != null && right == null) {
				replaceParentChild(left);
			}
			else if (left == null && right == null) {
				replaceParentChild(null);
			}
			
		}
    }
    BSTNode root;
    public BST() {}

    public void add(int value) {
        if (root == null)
            root = new BSTNode(value, null);
        else {
            root.add(value);
        }

    }

    public void delete(int value) {
		if (root != null)
			root.delete(value);
    }

    public boolean find(int value) {
		if (root != null)
			return root.search(value);
		return false;
	}
}
